### ISSUE TITLE
[BUG] Timeout of 30000ms exceeded while running fixture "browser" setup. (using Firefox)

### ISSUE THREAD

**Issue by eatonjl** [microsoft/playwright#14559]
**Context:**
- Playwright Version: 1.22
- Operating System: Ubuntu
- Node.js version: v16.15.0
- Browser: Firefox
- Extra: Run using TeamCity, on a Linux VM with a Docker TeamCity build agent and the Playwright Docker container

**Code Snippet**

Help us help you! Put down a short code snippet that illustrates your bug and
that we can run and debug locally. For example:

```javascript
import { test, expect, Page } from '@playwright/test';

test.beforeEach(async ({ page }) => {
  await page.goto('https://demo.playwright.dev/todomvc');
});

const TODO_ITEMS = [
  'buy some cheese',
  'feed the cat',
  'book a doctors appointment'
];

test.describe('New Todo', () => {
  test('should allow me to add todo items', async ({ page }) => {
    // Create 1st todo.
    await page.locator('.new-todo').fill(TODO_ITEMS[0]);
    await page.locator('.new-todo').press('Enter');

    // Make sure the list only has one todo item.
    await expect(page.locator('.view label')).toHaveText([
      TODO_ITEMS[0]
    ]);

    // Create 2nd todo.
    await page.locator('.new-todo').fill(TODO_ITEMS[1]);
    await page.locator('.new-todo').press('Enter');

    // Make sure the list now has two todo items.
    await expect(page.locator('.view label')).toHaveText([
      TODO_ITEMS[0],
      TODO_ITEMS[1]
    ]);

    await checkNumberOfTodosInLocalStorage(page, 2);
  });

  test('should clear text input field when an item is added', async ({ page }) => {
    // Create one todo item.
    await page.locator('.new-todo').fill(TODO_ITEMS[0]);
    await page.locator('.new-todo').press('Enter');

    // Check that input is empty.
    await expect(page.locator('.new-todo')).toBeEmpty();
    await checkNumberOfTodosInLocalStorage(page, 1);
  });

  test('should append new items to the bottom of the list', async ({ page }) => {
    // Create 3 items.
    await createDefaultTodos(page);

    // Check test using different methods.
    await expect(page.locator('.todo-count')).toHaveText('3 items left');
    await expect(page.locator('.todo-count')).toContainText('3');
    await expect(page.locator('.todo-count')).toHaveText(/3/);

    // Check all items in one call.
    await expect(page.locator('.view label')).toHaveText(TODO_ITEMS);
    await checkNumberOfTodosInLocalStorage(page, 3);
  });

  test('should show #main and #footer when items added', async ({ page }) => {
    await page.locator('.new-todo').fill(TODO_ITEMS[0]);
    await page.locator('.new-todo').press('Enter');

    await expect(page.locator('.main')).toBeVisible();
    await expect(page.locator('.footer')).toBeVisible();
    await checkNumberOfTodosInLocalStorage(page, 1);
  });
});

test.describe('Mark all as completed', () => {
  test.beforeEach(async ({ page }) => {
    await createDefaultTodos(page);
    await checkNumberOfTodosInLocalStorage(page, 3);
  });

  test.afterEach(async ({ page }) => {
    await checkNumberOfTodosInLocalStorage(page, 3);
  });

  test('should allow me to mark all items as completed', async ({ page }) => {
    // Complete all todos.
    await page.locator('.toggle-all').check();

    // Ensure all todos have 'completed' class.
    await expect(page.locator('.todo-list li')).toHaveClass(['completed', 'completed', 'completed']);
    await checkNumberOfCompletedTodosInLocalStorage(page, 3);
  });

  test('should allow me to clear the complete state of all items', async ({ page }) => {
    // Check and then immediately uncheck.
    await page.locator('.toggle-all').check();
    await page.locator('.toggle-all').uncheck();

    // Should be no completed classes.
    await expect(page.locator('.todo-list li')).toHaveClass(['', '', '']);
  });

  test('complete all checkbox should update state when items are completed / cleared', async ({ page }) => {
    const toggleAll = page.locator('.toggle-all');
    await toggleAll.check();
    await expect(toggleAll).toBeChecked();
    await checkNumberOfCompletedTodosInLocalStorage(page, 3);

    // Uncheck first todo.
    const firstTodo = page.locator('.todo-list li').nth(0);
    await firstTodo.locator('.toggle').uncheck();

    // Reuse toggleAll locator and make sure its not checked.
    await expect(toggleAll).not.toBeChecked();

    await firstTodo.locator('.toggle').check();
    await checkNumberOfCompletedTodosInLocalStorage(page, 3);

    // Assert the toggle all is checked again.
    await expect(toggleAll).toBeChecked();
  });
});

test.describe('Item', () => {

  test('should allow me to mark items as complete', async ({ page }) => {
    // Create two items.
    for (const item of TODO_ITEMS.slice(0, 2)) {
      await page.locator('.new-todo').fill(item);
      await page.locator('.new-todo').press('Enter');
    }

    // Check first item.
    const firstTodo = page.locator('.todo-list li').nth(0);
    await firstTodo.locator('.toggle').check();
    await expect(firstTodo).toHaveClass('completed');

    // Check second item.
    const secondTodo = page.locator('.todo-list li').nth(1);
    await expect(secondTodo).not.toHaveClass('completed');
    await secondTodo.locator('.toggle').check();

    // Assert completed class.
    await expect(firstTodo).toHaveClass('completed');
    await expect(secondTodo).toHaveClass('completed');
  });

  test('should allow me to un-mark items as complete', async ({ page }) => {
    // Create two items.
    for (const item of TODO_ITEMS.slice(0, 2)) {
      await page.locator('.new-todo').fill(item);
      await page.locator('.new-todo').press('Enter');
    }

    const firstTodo = page.locator('.todo-list li').nth(0);
    const secondTodo = page.locator('.todo-list li').nth(1);
    await firstTodo.locator('.toggle').check();
    await expect(firstTodo).toHaveClass('completed');
    await expect(secondTodo).not.toHaveClass('completed');
    await checkNumberOfCompletedTodosInLocalStorage(page, 1);

    await firstTodo.locator('.toggle').uncheck();
    await expect(firstTodo).not.toHaveClass('completed');
    await expect(secondTodo).not.toHaveClass('completed');
    await checkNumberOfCompletedTodosInLocalStorage(page, 0);
  });

  test('should allow me to edit an item', async ({ page }) => {
    await createDefaultTodos(page);

    const todoItems = page.locator('.todo-list li');
    const secondTodo = todoItems.nth(1);
    await secondTodo.dblclick();
    await expect(secondTodo.locator('.edit')).toHaveValue(TODO_ITEMS[1]);
    await secondTodo.locator('.edit').fill('buy some sausages');
    await secondTodo.locator('.edit').press('Enter');

    // Explicitly assert the new text value.
    await expect(todoItems).toHaveText([
      TODO_ITEMS[0],
      'buy some sausages',
      TODO_ITEMS[2]
    ]);
    await checkTodosInLocalStorage(page, 'buy some sausages');
  });
});

test.describe('Editing', () => {
  test.beforeEach(async ({ page }) => {
    await createDefaultTodos(page);
    await checkNumberOfTodosInLocalStorage(page, 3);
  });

  test('should hide other controls when editing', async ({ page }) => {
    const todoItem = page.locator('.todo-list li').nth(1);
    await todoItem.dblclick();
    await expect(todoItem.locator('.toggle')).not.toBeVisible();
    await expect(todoItem.locator('label')).not.toBeVisible();
    await checkNumberOfTodosInLocalStorage(page, 3);
  });

  test('should save edits on blur', async ({ page }) => {
    const todoItems = page.locator('.todo-list li');
    await todoItems.nth(1).dblclick();
    await todoItems.nth(1).locator('.edit').fill('buy some sausages');
    await todoItems.nth(1).locator('.edit').dispatchEvent('blur');

    await expect(todoItems).toHaveText([
      TODO_ITEMS[0],
      'buy some sausages',
      TODO_ITEMS[2],
    ]);
    await checkTodosInLocalStorage(page, 'buy some sausages');
  });

  test('should trim entered text', async ({ page }) => {
    const todoItems = page.locator('.todo-list li');
    await todoItems.nth(1).dblclick();
    await todoItems.nth(1).locator('.edit').fill('    buy some sausages    ');
    await todoItems.nth(1).locator('.edit').press('Enter');

    await expect(todoItems).toHaveText([
      TODO_ITEMS[0],
      'buy some sausages',
      TODO_ITEMS[2],
    ]);
    await checkTodosInLocalStorage(page, 'buy some sausages');
  });

  test('should remove the item if an empty text string was entered', async ({ page }) => {
    const todoItems = page.locator('.todo-list li');
    await todoItems.nth(1).dblclick();
    await todoItems.nth(1).locator('.edit').fill('');
    await todoItems.nth(1).locator('.edit').press('Enter');

    await expect(todoItems).toHaveText([
      TODO_ITEMS[0],
      TODO_ITEMS[2],
    ]);
  });

  test('should cancel edits on escape', async ({ page }) => {
    const todoItems = page.locator('.todo-list li');
    await todoItems.nth(1).dblclick();
    await todoItems.nth(1).locator('.edit').press('Escape');
    await expect(todoItems).toHaveText(TODO_ITEMS);
  });
});

test.describe('Counter', () => {
  test('should display the current number of todo items', async ({ page }) => {
    await page.locator('.new-todo').fill(TODO_ITEMS[0]);
    await page.locator('.new-todo').press('Enter');
    await expect(page.locator('.todo-count')).toContainText('1');

    await page.locator('.new-todo').fill(TODO_ITEMS[1]);
    await page.locator('.new-todo').press('Enter');
    await expect(page.locator('.todo-count')).toContainText('2');

    await checkNumberOfTodosInLocalStorage(page, 2);
  });
});

test.describe('Clear completed button', () => {
  test.beforeEach(async ({ page }) => {
    await createDefaultTodos(page);
  });

  test('should display the correct text', async ({ page }) => {
    await page.locator('.todo-list li .toggle').first().check();
    await expect(page.locator('.clear-completed')).toHaveText('Clear completed');
  });

  test('should remove completed items when clicked', async ({ page }) => {
    const todoItems = page.locator('.todo-list li');
    await todoItems.nth(1).locator('.toggle').check();
    await page.locator('.clear-completed').click();
    await expect(todoItems).toHaveCount(2);
    await expect(todoItems).toHaveText([TODO_ITEMS[0], TODO_ITEMS[2]]);
  });

  test('should be hidden when there are no items that are completed', async ({ page }) => {
    await page.locator('.todo-list li .toggle').first().check();
    await page.locator('.clear-completed').click();
    await expect(page.locator('.clear-completed')).toBeHidden();
  });
});

test.describe('Persistence', () => {
  test('should persist its data', async ({ page }) => {
    for (const item of TODO_ITEMS.slice(0, 2)) {
      await page.locator('.new-todo').fill(item);
      await page.locator('.new-todo').press('Enter');
    }

    const todoItems = page.locator('.todo-list li');
    await todoItems.nth(0).locator('.toggle').check();
    await expect(todoItems).toHaveText([TODO_ITEMS[0], TODO_ITEMS[1]]);
    await expect(todoItems).toHaveClass(['completed', '']);

    // Ensure there is 1 completed item.
    checkNumberOfCompletedTodosInLocalStorage(page, 1);

    // Now reload.
    await page.reload();
    await expect(todoItems).toHaveText([TODO_ITEMS[0], TODO_ITEMS[1]]);
    await expect(todoItems).toHaveClass(['completed', '']);
  });
});

test.describe('Routing', () => {
  test.beforeEach(async ({ page }) => {
    await createDefaultTodos(page);
    // make sure the app had a chance to save updated todos in storage
    // before navigating to a new view, otherwise the items can get lost :(
    // in some frameworks like Durandal
    await checkTodosInLocalStorage(page, TODO_ITEMS[0]);
  });

  test('should allow me to display active items', async ({ page }) => {
    await page.locator('.todo-list li .toggle').nth(1).check();
    await checkNumberOfCompletedTodosInLocalStorage(page, 1);
    await page.locator('.filters >> text=Active').click();
    await expect(page.locator('.todo-list li')).toHaveCount(2);
    await expect(page.locator('.todo-list li')).toHaveText([TODO_ITEMS[0], TODO_ITEMS[2]]);
  });

  test('should respect the back button', async ({ page }) => {
    await page.locator('.todo-list li .toggle').nth(1).check();
    await checkNumberOfCompletedTodosInLocalStorage(page, 1);

    await test.step('Showing all items', async () => {
      await page.locator('.filters >> text=All').click();
      await expect(page.locator('.todo-list li')).toHaveCount(3);
    });

    await test.step('Showing active items', async () => {
      await page.locator('.filters >> text=Active').click();
    });

    await test.step('Showing completed items', async () => {
      await page.locator('.filters >> text=Completed').click();
    });

    await expect(page.locator('.todo-list li')).toHaveCount(1);
    await page.goBack();
    await expect(page.locator('.todo-list li')).toHaveCount(2);
    await page.goBack();
    await expect(page.locator('.todo-list li')).toHaveCount(3);
  });

  test('should allow me to display completed items', async ({ page }) => {
    await page.locator('.todo-list li .toggle').nth(1).check();
    await checkNumberOfCompletedTodosInLocalStorage(page, 1);
    await page.locator('.filters >> text=Completed').click();
    await expect(page.locator('.todo-list li')).toHaveCount(1);
  });

  test('should allow me to display all items', async ({ page }) => {
    await page.locator('.todo-list li .toggle').nth(1).check();
    await checkNumberOfCompletedTodosInLocalStorage(page, 1);
    await page.locator('.filters >> text=Active').click();
    await page.locator('.filters >> text=Completed').click();
    await page.locator('.filters >> text=All').click();
    await expect(page.locator('.todo-list li')).toHaveCount(3);
  });

  test('should highlight the currently applied filter', async ({ page }) => {
    await expect(page.locator('.filters >> text=All')).toHaveClass('selected');
    await page.locator('.filters >> text=Active').click();
    // Page change - active items.
    await expect(page.locator('.filters >> text=Active')).toHaveClass('selected');
    await page.locator('.filters >> text=Completed').click();
    // Page change - completed items.
    await expect(page.locator('.filters >> text=Completed')).toHaveClass('selected');
  });
});

async function createDefaultTodos(page: Page) {
  for (const item of TODO_ITEMS) {
    await page.locator('.new-todo').fill(item);
    await page.locator('.new-todo').press('Enter');
  }
}

async function checkNumberOfTodosInLocalStorage(page: Page, expected: number) {
  return await page.waitForFunction(e => {
    return JSON.parse(localStorage['react-todos']).length === e;
  }, expected);
}

async function checkNumberOfCompletedTodosInLocalStorage(page: Page, expected: number) {
  return await page.waitForFunction(e => {
    return JSON.parse(localStorage['react-todos']).filter((todo: any) => todo.completed).length === e;
  }, expected);
}

async function checkTodosInLocalStorage(page: Page, title: string) {
  return await page.waitForFunction(t => {
    return JSON.parse(localStorage['react-todos']).map((todo: any) => todo.title).includes(t);
  }, title);
}
```

**Describe the bug**

I have other code I am using for this, but I was able to reproduce the issue with the default example tests provided here that Playwright provides when downloading it. I am running my tests on TeamCity. I have a Linux VM with a TeamCity build agent Docker container. I am also using the Docker container from Playwright to run the tests in. So far, the tests run fine in Chromium and Safari. However, I get the following error when trying to run them in Firefox: 
**Timeout of 30000ms exceeded while running fixture "browser" setup.
    at /opt/buildagent/work/6f7a18f2c8f56a5c/PolicyTechAutomationTests/tests/smoketestshould.spec.ts:94:6** 
That line in my code (which is different than the code pasted here) is: **test.beforeAll(async ({ browser }) => {** 
I am able to run these tests using Firefox on my local environment without issue. It is almost as if the Playwright Docker container is acting like Firefox does not exist.

**Comment by eatonjl** [comment]
Another thing to note is that I am using environment variables. I tried with an invalid browser as well (e.g. "firefoxas" instead of "firefox"). This makes me think it may not recognize the Firefox variable for the project for some reason.

**Comment by KenjiTakahashi** [comment]
I'm also seeing this problem, but only when running inside Docker container. Running the same tests on the same environment, but directly on the machine, works fine.
Extending timeout doesn't help anything.
I can see some `firefox` instances getting spawned, but they don't seem work as expected.

**Comment by gegoncalves** [comment]
Having the same problem described by @KenjiTakahashi

**Comment by aslushnikov** [comment]
Hey folks, I tried the following:

1. kicked docker container and installed Playwright Test inside:
```bash
aslushnikov:~/prog/playwright(main)$ docker run --rm -it mcr.microsoft.com/playwright:v1.22.2-focal /bin/bash
root@91feb1a105a8:/# cd
root@91feb1a105a8:~# npm init -y && npm i @playwright/test
```

2. Copied the test suite from https://github.com/microsoft/playwright/issues/14559#issue-1256944229 as `a.spec.ts` inside docker container
3. Run the tests with `npx playwright test --browser firefox --repeat-each=10` inside docker

I ran the tests a few times and they worked just fine. Can you help me reproduce this?

**Comment by eatonjl** [comment]
Good to know, @KenjiTakahashi. As for the config, I can include mine here. I am using environment variables to dictate the browser and device. 
[Link To Config Code](https://carbon.now.sh/?bg=rgba%28171%2C+184%2C+195%2C+1%29&t=cobalt&wt=none&l=application%2Ftypescript&width=680&ds=true&dsyoff=20px&dsblur=68px&wc=true&wa=true&pv=56px&ph=56px&ln=false&fl=1&fm=Hack&fs=14px&lh=133%25&si=false&es=1x&wm=false&code=import%2520type%2520%257B%2520PlaywrightTestConfig%2520%257D%2520from%2520%27%2540playwright%252Ftest%27%253B%250Aimport%2520%257B%2520devices%2520%257D%2520from%2520%27%2540playwright%252Ftest%27%253B%250Aimport%2520dotenv%2520from%2520%27dotenv%27%253B%250A%250A%252F**%250A%2520*%2520Read%2520environment%2520variables%2520from%2520file.%250A%2520*%2520https%253A%252F%252Fgithub.com%252Fmotdotla%252Fdotenv%250A%2520*%252F%250Adotenv.config%28%29%253B%250A%250A%252F**%250A%2520*%2520See%2520https%253A%252F%252Fplaywright.dev%252Fdocs%252Ftest-configuration.%250A%2520*%252F%250Aconst%2520config%253A%2520PlaywrightTestConfig%2520%253D%2520%257B%250A%2509testDir%253A%2520%27.%252Ftests%27%252C%250A%2509%252F*%2520Maximum%2520time%2520one%2520test%2520can%2520run%2520for.%2520*%252F%250A%2509timeout%253A%252030%2520*%25201000%252C%250A%2509expect%253A%2520%257B%250A%2509%2509%252F**%250A%2509%2509%2520*%2520Maximum%2520time%2520expect%28%29%2520should%2520wait%2520for%2520the%2520condition%2520to%2520be%2520met.%250A%2509%2509%2520*%2520For%2520example%2520in%2520%2560await%2520expect%28locator%29.toHaveText%28%29%253B%2560%250A%2509%2509%2520*%252F%250A%2509%2509timeout%253A%252010000%252C%250A%2509%257D%252C%250A%2509%252F%252F%2520Tell%2520all%2520tests%2520to%2520load%2520signed-in%2520state%2520from%2520%27storageState.json%27.%250A%2509globalSetup%253A%2520require.resolve%28%27.%252Fglobal-setup%27%29%252C%250A%2509%252F*%2520Fail%2520the%2520build%2520on%2520CI%2520if%2520you%2520accidentally%2520left%2520test.only%2520in%2520the%2520source%2520code.%2520*%252F%250A%2509forbidOnly%253A%2520%21%21process.env.CI%252C%250A%2509%252F*%2520Retry%2520on%2520CI%2520only%2520*%252F%250A%2509retries%253A%2520process.env.CI%2520%253F%25202%2520%253A%25200%252C%250A%2509%252F*%2520Opt%2520out%2520of%2520parallel%2520tests%2520on%2520CI.%2520*%252F%250A%2509workers%253A%25201%252C%250A%2509%252F*%2520Reporter%2520to%2520use.%2520See%2520https%253A%252F%252Fplaywright.dev%252Fdocs%252Ftest-reporters%2520*%252F%250A%2509%252F%252Freporter%253A%2520process.env.CI%2520%253F%2520%27dot%27%2520%253A%2520%27list%27%252C%250A%2509reporter%253A%2520process.env.CI%250A%2509%2509%253F%2520%255B%250A%2509%2509%2509%2509%255B%27list%27%255D%252C%250A%2509%2509%2509%2509%255B%250A%2509%2509%2509%2509%2509%27playwright-teamcity-reporter%27%252C%250A%2509%2509%2509%2509%2509%257B%2520testMetadataArtifacts%253A%2520%27test-results%27%252C%2520logConfig%253A%2520false%2520%257D%252C%250A%2509%2509%2509%2509%255D%252C%250A%2509%2509%2520%2520%255D%250A%2509%2509%253A%2520%27list%27%252C%250A%2509%252F%252F%2520reporter%253A%2520%255B%250A%2509%252F%252F%2520%2509%255B%27list%27%255D%252C%250A%2509%252F%252F%2520%2509%255B%27playwright-teamcity-reporter%27%252C%2520%257B%2520testMetadataArtifacts%253A%2520%27test-results%27%2520%257D%255D%252C%250A%2509%252F%252F%2520%255D%252C%250A%2509%252F*%2520Shared%2520settings%2520for%2520all%2520the%2520projects%2520below.%2520See%2520https%253A%252F%252Fplaywright.dev%252Fdocs%252Fapi%252Fclass-testoptions.%2520*%252F%250A%2509use%253A%2520%257B%250A%2509%2509ignoreHTTPSErrors%253A%2520true%252C%250A%2509%2509storageState%253A%2520%27storageState.json%27%252C%250A%2509%2509%252F*%2520Maximum%2520time%2520each%2520action%2520such%2520as%2520%2560click%28%29%2560%2520can%2520take.%2520Defaults%2520to%25200%2520%28no%2520limit%29.%2520*%252F%250A%2509%2509actionTimeout%253A%25200%252C%250A%2509%2509%252F*%2520Base%2520URL%2520to%2520use%2520in%2520actions%2520like%2520%2560await%2520page.goto%28%27%252F%27%29%2560.%2520*%252F%250A%2509%2509baseURL%253A%2520process.env.BASEURL%252C%250A%250A%2509%2509%252F*%2520Collect%2520trace%2520when%2520retrying%2520the%2520failed%2520test.%2520See%2520v%2520*%252F%250A%2509%2509trace%253A%2520%27retain-on-failure%27%252C%250A%2509%257D%252C%250A%250A%2509%252F*%2520Configure%2520projects%2520for%2520major%2520browsers%2520*%252F%250A%2509projects%253A%2520%255B%250A%2509%2509%257B%250A%2509%2509%2509name%253A%2520process.env.BROWSER_NAME%252C%250A%2509%2509%2509use%253A%2520%257B%250A%2509%2509%2509%2509...devices%255Bprocess.env.DEVICE%255D%252C%250A%2509%2509%2509%257D%252C%250A%2509%2509%257D%252C%250A%250A%2509%2509%252F%252F%2520%257B%250A%2509%2509%252F%252F)

**Comment by eatonjl** [comment]
I tried a change to see how it would react. I removed the environment variables, had three projects (one for each browser types) and indicated the project in the TeamCity settings. The results were a little different, but still not good. It ran the first three tests before ignoring the rest (whereas before it only ran the first one). However, the failure message for all three was the same as the failure message for the one failed test before. So I believe that rules out the environment variables causing the issue.

**Comment by KenjiTakahashi** [comment]
It looks to me like it's somehow related to tests code and/or `node_modules` being inside a mounted volume.
What I did:
1. `npm init -y playwright@latest pwtest` (leave all at default)
2. `cd pwtest`
3. `docker run --rm -it -v $(pwd):/code/ mcr.microsoft.com/playwright:v1.22.2-focal /bin/bash`
4. [Inside container] `cd /code/ && npx playwright test --project firefox`
5. No profit :-(.

**Comment by jkilada** [comment]
I am also trying to run tests in a CI environment with the docker container bind mounting the source directory, and experiencing this same issue.

**Comment by eatonjl** [comment]
@aslushnikov Do you feel you have enough from these comments to reproduce the issue? Please let me know if you need more. @KenjiTakahashi and @jkilada, thank you for your help in figuring out how to reproduce.

**Comment by aslushnikov** [comment]
**Everybody:** I see you all mount the source folder inside the container. Do you run `npm i` inside the container as well? If not, could you please try running `npm i` and check if the issue goes away?

**Comment by gegoncalves** [comment]
> **Everybody:** I see you all mount the source folder inside the container. Do you run `npm i` inside the container as well? If not, could you please try running `npm i` and check if the issue goes away?

We did it and it is working ok. By the way: we were at the 1.21 - we changed to the 1.22 and `npm i`

@aslushnikov

**Comment by eatonjl** [comment]
I am not sure I understand (I am pretty new to this ). In Teamcity I have to commands for the build step. The first is **npm ci** and the second is **npx playwright test tests/smoketestshould.spec.ts --project=firefox**
Are you saying I should replace **npm ci** with **npm I** in the build step in TeamCity?
Also, would your recommend we not mount the source folder inside the container?

**Comment by aslushnikov** [comment]
> Are you saying I should replace npm ci with npm I in the build step in TeamCity?

@eatonjl No, `npm ci` works well. 

> Also, would your recommend we not mount the source folder inside the container?

This is a fine practice, however all the build artifacts have to be re-built inside container. The `npm i` / `npm ci` is the command to build artifacts inside.

**Comment by eatonjl** [comment]
@aslushnikov that is good to know. So, should what I am doing work, or am I also supposed to put an `npm I` somewhere else as well?

**Comment by aslushnikov** [comment]
@eatonjl it all should work, at least from the Playwright side.

> I think the one thing missing from what you tried (unless I am reading it wrong) would be the absence of TeamCity.

I unfortunately don't have much experience with TeamCity, so I won't be able to suggest much there. But it looks like something is wrong in the TeamCity configuration.

**Comment by eatonjl** [comment]
Earlier I included the pull and run commands used in TeamCity. According to what you see below, do you think I would need to add `npm I` to the run command after it pulls the Playwright Docker image? 
`/bin/sh -c "docker pull mcr.microsoft.com/playwright:v1.22.0-focal 
&& 
. /opt/buildagent/temp/agentTmp/docker-wrapper-4487303948225419876.sh 
&& 
docker run --rm -w /opt/buildagent/work/6f7a18f2c8f56a5c/PolicyTechAutomationTests --label jetbrains.teamcity.buildId=107566 --network host --ipc=host -v "/opt/buildagent/lib:/opt/buildagent/lib:ro" -v "/opt/buildagent/tools:/opt/buildagent/tools:ro" -v "/opt/buildagent/plugins:/opt/buildagent/plugins:ro" -v "/opt/buildagent/work/6f7a18f2c8f56a5c:/opt/buildagent/work/6f7a18f2c8f56a5c" -v "/opt/buildagent/temp/agentTmp:/opt/buildagent/temp/agentTmp" -v "/opt/buildagent/temp/buildTmp:/opt/buildagent/temp/buildTmp" -v "/opt/buildagent/system:/opt/buildagent/system" --env-file /opt/buildagent/temp/agentTmp/docker-wrapper-14098506302688146999.envList --entrypoint /bin/sh "mcr.microsoft.com/playwright:v1.22.0-focal" /opt/buildagent/temp/agentTmp/docker-shell-script-6445093165381205156.sh"`

**Comment by aslushnikov** [comment]
>  do you think I would need to add npm I to the run command after it pulls the Playwright Docker image?

@eatonjl The `npm i` has to be inside container; the command you showed me runs outside the container. So adding `npm i` won't change anything.

Could you please show the steps you run inside the container?

**Comment by eatonjl** [comment]
@aslushnikov, well I thought I just did. Like I said, I am pretty new to this. There are three parts in TeamCity that may be relevant here. The first is the shell script. That is where I run the `npm ci` and `npx playwright test tests/smoketestshould.spec.ts --project=firefox` commands. Then there is the "Run step within Docker container:" area. This is where the "mcr.microsoft.com/playwright:v1.22.0-focal" line gets put. The third is the "Additional docker run arguments:" area. In there, I can type additions to the docker run command (such as --network host or --ipc=host).

When the build runs, it gives this output: 
**Running step within Docker container mcr.microsoft.com/playwright:v1.22.0-focal
Starting: /bin/sh -c "docker pull mcr.microsoft.com/playwright:v1.22.0-focal && . /opt/buildagent/temp/agentTmp/docker-wrapper-11693806245354486042.sh && docker run --rm -w /opt/buildagent/work/6f7a18f2c8f56a5c/PolicyTechAutomationTests --label jetbrains.teamcity.buildId=107755 --network host --ipc=host -v "/opt/buildagent/lib:/opt/buildagent/lib:ro" -v "/opt/buildagent/tools:/opt/buildagent/tools:ro" -v "/opt/buildagent/plugins:/opt/buildagent/plugins:ro" -v "/opt/buildagent/work/6f7a18f2c8f56a5c:/opt/buildagent/work/6f7a18f2c8f56a5c" -v "/opt/buildagent/temp/agentTmp:/opt/buildagent/temp/agentTmp" -v "/opt/buildagent/temp/buildTmp:/opt/buildagent/temp/buildTmp" -v "/opt/buildagent/system:/opt/buildagent/system" --env-file /opt/buildagent/temp/agentTmp/docker-wrapper-9206620540927956518.envList --entrypoint /bin/sh "mcr.microsoft.com/playwright:v1.22.0-focal" /opt/buildagent/temp/agentTmp/docker-shell-script-8642703566818739957.sh"
in directory: /opt/buildagent/work/6f7a18f2c8f56a5c/PolicyTechAutomationTests
v1.22.0-focal: Pulling from playwright
Digest: sha256:4acad6159e3b6e790b11d7c62af7008d843f2d9f687868e40e65a2871d797b38
Status: Image is up to date for mcr.microsoft.com/playwright:v1.22.0-focal
mcr.microsoft.com/playwright:v1.22.0-focal**

After that, assume it runs the `npm ci` and `npx playwright test tests/smoketestshould.spec.ts --project=firefox` in the mcr.microsoft.com/playwright:v1.22.0-focal container it set up.

**Comment by eatonjl** [comment]
According to TeamCity documentation, it runs those two commands in the Docker container:
"Run step within Docker container | Specify a Docker image name as stated in Docker Hub. TeamCity will start a container from the specified image and will try to run this build step within this container."
https://www.jetbrains.com/help/teamcity/2022.04/docker-wrapper.html#Docker+Settings

**Comment by KenjiTakahashi** [comment]
@aslushnikov Installing from inside the container does not fix the issue for me.
I did some further experiments.
Using the same example project code from `npm init`, but removed `node_modules` and `package*json`.

`docker run --rm -it -v $(pwd):/code/t/ mcr.microsoft.com/playwright:v1.22.2-focal /bin/bash`

Then inside the container:
1. `cd /code/`
2. `npm i @playwright/test`
3. `npx playwright test -c /code/t/playwright.config.ts --project firefox` -> This works fine.
4. `cd /t/`
5. `npx playwright test -c /code/t/playwright.config.ts --project firefox` -> This fails.

Then I did the "opposite" (in new container):
1. `cd /code/t/`
2. `npm i @playwright/test`
3. `npx --prefix /code/t/ playwright test -c /code/t/playwright.config.ts --project firefox` -> This fails.
4. `cd ..`
5. `npx --prefix /code/t/ playwright test -c /code/t/playwright.config.ts --project firefox` -> This works.

So it looks like the trigger is `cwd` being on a bind mount. Doesn't matter if the code or deps are also bind mounted or not.
But why that is, I can't really tell. And why it only fails for Firefox? Sth wrong with how Firefox handles files in bind mounted dirs?

**Comment by jkilada** [comment]
Thanks @KenjiTakahashi for your research. I was able to resolve the issue in a similar manner by ensuring cwd was outside of the bind mount before invoking the tests.

**Comment by aslushnikov** [comment]
> Here is an image showing how these are entered into the TeamCity interface: https://ibb.co/nDwwGs7

@eatonjl thank you for sharing the screenshot, your setup seems legit to me. Can you try the @KenjiTakahashi workaround?

**Comment by aslushnikov** [comment]
> So it looks like the trigger is cwd being on a bind mount. Doesn't matter if the code or deps are also bind mounted or not.
But why that is, I can't really tell. And why it only fails for Firefox? Sth wrong with how Firefox handles files in bind mounted dirs?

@KenjiTakahashi Thank you for the investigation! I didn't find anything regarding this on the BugZilla. I wonder if you can reproduce the same behavior without Playwright, using plain `firefox -headless -screenshot` commands?

**Comment by eatonjl** [comment]
> > Here is an image showing how these are entered into the TeamCity interface: https://ibb.co/nDwwGs7
> 
> @eatonjl thank you for sharing the screenshot, your setup seems legit to me. Can you try the @KenjiTakahashi workaround?

@aslushnikov I should be able to get some time to try that today. I will let you know what I find.

**Comment by KenjiTakahashi** [comment]
I've tried running `firefox -headless -screenshot` on the same setup and it worked fine every time. Looks like something else going on.

**Comment by eatonjl** [comment]
@aslushnikov OK, I was able to take a look at it. I had a developer on my team who knew a little more than me try out the workaround. Basically, in TeamCity we removed the working directory. In the shell script we cd'd into the working directory and did the install command. Then, we backed out of the directory and execute the run command, this time specifying where the config is in the working directory folder. It works! Not sure why it does that, though. Definitely seems buggy to me. It would be nice to get that fixed so we can go back to having a working directory and two simple build/run commands.

**Comment by aslushnikov** [comment]
**everybody**: Lots of things have landed to Firefox to address this and other issues. Could you please try it out with canary version?

```bash
npm i @playwright/test@next
npx playwright install firefox
```

**Comment by eatonjl** [comment]
Sure! I will hopefully be able to try today (busy day). If not, I can try it tomorrow.

**Comment by skysantoroa** [comment]
Getting the same issue when binding a volume containing the `node_modules` folder. The problem will disappear if the volume is mounted only in a folder that contains the `.spec.*` files.

**Comment by eatonjl** [comment]
@aslushnikov Even thought the suggestion you gave last didn't work, we have it working now. We had an issues with running the line for downloading the browsers (_npx playwright install --with-deps_). The error was: "Password: su: Authentication failure"
Apparently, when the playwright script is run through npx/npm it assumes the user role of whoever owns the directory from where it is run. In this case TeamCity checks out the code with the BuildAgent user permissions. In the end we just had to add a single extra line to the build steps: _chown root ._
With that addition, it not only solved the authentication failure but also the need to do the workaround I listed above. I am not sure if we are supposed to need _chown root ._ before _npm ci_ and _npx playwright install --with-deps_ to get it to work. The documentation does not specify needing it. If we do though, I don't see this as needing to stay open. If we shouldn't need to use _chown root ._, then it still remains an issue.

**Comment by eatonjl** [comment]
@aslushnikov I was able to reduce the lines of shell script I was using. I realized that, since I am using the Playwright Docker image, I had a lot of redundant script lines. The lines I was able to reduce it down to was this:
```
chown root .
npx playwright test tests/smoketestshould --project=firefox
```
I no longer have the issue, but I am not sure why I need to use "chown root ." as the other browsers don't require it. I have that issue being worked on [here](https://github.com/microsoft/playwright/issues/16099)

I am good with this issue being closed unless others who still have the issue require further help with it.
